### PR TITLE
Harmonic number

### PR DESCRIPTION
**Pull request by 0x00b1**
## Harmonic number

```Python
harmonic_number(n, *, out=None) -> Tensor
```

$n^{\textup{th}}$ harmonic number. 

Parameters:

n ([Tensor](https://pytorch.org/docs/stable/tensors.html#torch.Tensor)) – the input tensor. `n` must have an integral [dtype](https://pytorch.org/docs/stable/tensor_attributes.html#torch-dtype).

Keyword Arguments:

out ([Tensor](https://pytorch.org/docs/stable/tensors.html#torch.Tensor), optional) – the output tensor.

### Checklist

- [x] structured kernel (or have a good reason not to be)
- [x] jiterated
- [x] documented via docstring 
- [x] updated `docs/source/special.rst`
- [x] exposed on the `torch.special` namespace
- [x] function and out variant (2*N new entries there) added to native_functions.yaml
- [x] tested via an OpInfo that includes a ref
- [x] functions for the cpp API in csrc/api/include (optional but should be consistent)
- [x] derivatives.yaml contain the right enty for each function (N new entries)
- [x] nothing is added to the public API allowlist
- [x] Sanity check the c++ headers to only have minimal includes
- [x] per-operator headers are being used and no large includes like ATen.h or Function.h

### CODE REVIEW

**Comment by 0x00b1** [comment]
@albanD Would you mind taking a look at these failures? They are non-obvious to me.

File: aten/src/ATen/native/special_functions/detail/numeric.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <c10/util/complex.h>
+
+namespace at {
```
Comment (albanD) [inline aten/src/ATen/native/special_functions/detail/numeric.h:5]: These are duplicated with the other PR and will go away once it is merged right?

File: torch/special/__init__.py [hunk 2]
```diff
@@ -47,12 +48,12 @@
     'polygamma',
     'psi',
     'round',
+    'scaled_modified_bessel_k0',
```
Comment (albanD) [inline torch/special/__init__.py:51]: nit: there are a lot of shuffling of other names (in c++ code as well btw), reducing these would be nice to reduce noise.

File: torch/testing/_internal/opinfo/definitions/special.py [hunk 3]
```diff
@@ -647,6 +658,20 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
         ref=lambda x: scipy.special.spherical_jn(0, x) if TEST_SCIPY else None,
         supports_autograd=False,
     ),
+    UnaryUfuncInfo(
+        "special.harmonic_number",
+        decorators=(
+            toleranceOverride(
+                {
+                    torch.float32: tol(atol=1e-05, rtol=1e-05),
+                    torch.float64: tol(atol=1e-05, rtol=1e-05),
```
Comment (albanD) [inline torch/testing/_internal/opinfo/definitions/special.py:667]: This is very high for float64 no?

**Review by albanD — COMMENTED** [review]
Sounds good

File: torch/testing/_internal/opinfo/definitions/special.py (hunk 3 above)
Comment (0x00b1) [reply torch/testing/_internal/opinfo/definitions/special.py:667]: This is because of my less than ideal ref.

File: torch/special/__init__.py (hunk 2 above)
Comment (0x00b1) [reply torch/special/__init__.py]: OK! I removed all of them.
